gcp_variant_transforms/libs/infer_headers_util.py

```diff
@@ -0,0 +1,360 @@
+      raise ValueError(error)
+
+  resolver = vcf_field_conflict_resolver.FieldConflictResolver(
+      resolve_always=True)
+  for field in annotation_fields_to_infer:
+    if field not in variant.info:
+      continue
+    annotation_names = annotation_parser.extract_annotation_names(
+        defined_headers.infos[field][_HeaderKeyConstants.DESC])
+    # First element (ALT) is ignored, since its type is hard-coded as string
+    annotation_values = [annotation_parser.extract_annotation_list_with_alt(
+        annotation)[1:] for annotation in variant.info[field]]
+    _check_annotation_lists_lengths(annotation_names, annotation_values)
+    annotation_values = zip(*annotation_values)
+    for name, values in zip(annotation_names, annotation_values):
+      variant_merged_type = None
+      for v in values:
+        if not v:
+          continue
+        variant_merged_type = resolver.resolve_attribute_conflict(
+            _HeaderKeyConstants.TYPE,
+            variant_merged_type,
+            _get_field_type(v))
+        if variant_merged_type == _HeaderTypeConstants.STRING:
+          break
+      key_id = annotation_parser.get_inferred_annotation_type_header_key(
+          field, name)
+      infos[key_id] = Info(key_id,
+                           1,  # field count
+                           variant_merged_type,
+                           ('Inferred type field for annotation {}.'.format(
+                               name)),
+                           '',  # UNKNOWN_SOURCE
+                           '')  # UNKNOWN_VERSION
+
+def infer_info_fields(
+    variant,
+    defined_headers,
+    infer_headers=False,  # type: bool
+    annotation_fields_to_infer=None  # type: Optional[List[str]]
+    ):
+  """Returns inferred info fields.
+
+  Up to three types of info fields are inferred:
+
+  if `infer_headers` is True:
+  - The info fields are undefined in the headers.
+  - The info fields' definitions provided by the header does not match the
+    field value.
+  if `infer_annotation_types` is True:
+  - Fields containing type information of corresponding annotation Info
+    fields.
+
+  Args:
+    variant: variant object
+    defined_headers: header fields defined in header section of VCF files.
+    infer_headers: If true, header fields are inferred from variant data.
+    annotation_fields_to_infer: list of info fields treated as annotation
+        fields (e.g. ['CSQ', 'CSQ_VT']).
+  Returns:
+    infos: dict of (info_key, `Info`) for any info field in
+      `variant` that is not defined in the header or the definition mismatches
+      the field values.
+  """
+  infos = {}
+  if infer_headers:
+    _infer_standard_info_fields(variant, infos, defined_headers)
+  if annotation_fields_to_infer:
+    _infer_annotation_type_info_fields(
+        variant, infos, defined_headers, annotation_fields_to_infer)
+  return infos
+
+def infer_format_fields(variant, defined_headers):
```

Comment:
In general, I prefer to put the public methods at the beginning of one script, rather than scrolling all the way down to find out what this file can provide. If you agree, let's stick to the same rule. Of course, there is no wrong/right for the other way. Just let me know your preference.